### DISCUSSION TITLE
No way to build app with supabaseServer or supabaseClient - Dynamic server usage: Page couldn't be rendered statically because it used `cookies` - or You're importing a component that needs next/headers

### DISCUSSION THREAD

**Discussion by nicitaacom** [supabase/supabase#18086] [Contribute to Supabase]
# Bug report

Dynamic server usage: Page couldn't be rendered statically because it used `cookies`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error

- [x] I confirm this is a bug with Supabase, not with my own application.  - yes because I found this issue on next js github ant it was closed
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com).  - yes because to reproduce this I use docs
- 
## Describe the bug

When you use supabaseServer you usually get this error
```
Dynamic server usage: Page couldn't be rendered statically because it used `cookies`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error
```

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

codesandbox - https://codesandbox.io/p/github/nicitaacom/auth-cookie-supabase-willbedeleted/auth-form-next
github repository - https://github.com/nicitaacom/auth-cookie-supabase-willbedeleted/blob/main/app/components/Navbar/components/UserIcon.tsx
(this file that trigger the error)
1. Open codesandbox
2. Go to app/components/Navbar/Navbar.tsx
3. Check that I'm doing something wrong (or it something wrong with library)

## Expected behavior

I got an error but should be no error

## System information

```json
    "next": "^13.5.4",
    "@supabase/auth-helpers-nextjs": "^0.8.1",
    "@supabase/supabase-js": "^2.38.0",
```

- Version of Node.js: 20.8.0

## Additional context

I use supabaseServer like this

**Navbar.tsx**
```
import supabaseServer from "@/utils/supabaseServer"
import { OpenAuthModalButton, UserIcon } from "./components"

export default async function Navbar() {
  const user = await supabaseServer.auth.getUser()

  return (
    <nav className="px-8 py-4 flex justify-between items-center gap-x-4">
      <OpenAuthModalButton />
      <UserIcon user={user} />
    </nav>
  )
}

```

**UserIcon.tsx**
```tsx
import Image from "next/image"

import { BiUserCircle } from "react-icons/bi"
import { UserResponse } from "@supabase/supabase-js"

interface UserIconProps {
  user: UserResponse
}

export default function UserIcon({ user }: UserIconProps) {
  return (
    <>
      {!user.data.user?.user_metadata.avatar ? (
        <BiUserCircle className="cursor-pointer" size={32} />
      ) : user.data.user?.user_metadata.avatar ? (
        <Image
          className="rounded-full cursor-pointer"
          src={user.data.user?.user_metadata.avatar}
          alt="profile_picture_url"
          width={32}
          height={32}
        />
      ) : (
        <Image
          className="rounded-full cursor-pointer"
          src="/placeholder.jpg"
          alt="profile_picture_url"
          width={32}
          height={32}
        />
      )}
    </>
  )
}

```

**supabaseServer.ts**
```ts
import { createServerComponentClient } from "@supabase/auth-helpers-nextjs"
import { cookies } from "next/headers"
import { Database } from "@/interfaces/types_db"

const supabaseServer = createServerComponentClient<Database>({ cookies })

export default supabaseServer
```

**middleware.ts**
```ts
import { createMiddlewareClient } from "@supabase/auth-helpers-nextjs";
import { NextRequest, NextResponse } from "next/server";

export async function middleware(req:NextRequest) {
	const res = NextResponse.next();
	const supabase = createMiddlewareClient({ req, res });
	await supabase.auth.getSession();
	return res;
}
```

I tried figure out this isse any possible way I write this code in some other ways but it still doesn't work
Please don't leave me alone with this issue - I'm stucked and really need your help

**Comment by saltcod** [comment]
Converting to a Discussion. 
Have you followed the docs here? 
https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs

**Reply by nicitaacom** [reply]
>> Have you followed the docs here?
https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs

I used this docs when I first time wanted to write auth
I don't like use <Auth/> because its no way to customize it as I want
Also I want insert rows in tables when user create account (I know it possible to do with triggers - if you know how let me know)

**Comment by nicitaacom** [comment]
**Fixed** - https://github.com/vercel/next.js/issues/56630
**By** - https://github.com/Flushey
